File: app/src/main/java/nl/tudelft/cs4160/trustchain_android/chainExplorer/ChainExplorerActivity.java
package nl.tudelft.cs4160.trustchain_android.chainExplorer;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.GridLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.ProgressBar;

import java.security.KeyPair;

import nl.tudelft.cs4160.trustchain_android.R;
import nl.tudelft.cs4160.trustchain_android.Util.Key;
import nl.tudelft.cs4160.trustchain_android.database.TrustChainDBHelper;

import static android.view.Gravity.CENTER;


public class ChainExplorerActivity extends AppCompatActivity {
    TrustChainDBHelper dbHelper;
    ChainExplorerAdapter adapter;
    ListView blocksList;

    static final String TAG = "ChainExplorerActivity";


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chain_explorer);
        blocksList = (ListView) findViewById(R.id.blocks_list);

        // Create a progress bar to display while the list loads
        ProgressBar progressBar = new ProgressBar(this);
        progressBar.setLayoutParams(new LinearLayout.LayoutParams(GridLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT, CENTER));
        progressBar.setIndeterminate(true);
        blocksList.setEmptyView(progressBar);

        // Must add the progress bar to the root of the layout
        ViewGroup root = (ViewGroup) findViewById(android.R.id.content);
        root.addView(progressBar);

        init();
    }

    private void init() {
        dbHelper = new TrustChainDBHelper(this);
        KeyPair kp = Key.loadKeys(getApplicationContext());
        try {
            adapter = new ChainExplorerAdapter(this, dbHelper.getAllBlocks(), kp.getPublic().getEncoded());
            blocksList.setAdapter(adapter);
        } catch (Exception e) {
            e.printStackTrace();
        }

        blocksList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                LinearLayout expandedItem = (LinearLayout) view.findViewById(R.id.expanded_item);
                ImageView expandArrow = (ImageView) view.findViewById(R.id.expand_arrow);

                // Expand the item when it is clicked
                if(expandedItem.getVisibility() == View.GONE) {
                    expandedItem.setVisibility(View.VISIBLE);
                    Log.v(TAG, "Item height: " + expandedItem.getHeight());
                    expandArrow.setImageDrawable(getDrawable(R.drawable.ic_expand_less_black_24dp));
                } else {
                    expandedItem.setVisibility(View.GONE);
                    expandArrow.setImageDrawable(getDrawable(R.drawable.ic_expand_more_black_24dp));
                }
            }
        });
    }

}

File: app/src/main/java/nl/tudelft/cs4160/trustchain_android/chainExplorer/ChainExplorerActivity.kt
package nl.tudelft.cs4160.trustchain_android.chainExplorer


import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.Gravity.CENTER
import android.view.View
import android.view.ViewGroup
import android.widget.*
import kotlinx.android.synthetic.main.activity_chain_explorer.*
import kotlinx.android.synthetic.main.item_trustchainblock.view.*
import nl.tudelft.cs4160.trustchain_android.R
import nl.tudelft.cs4160.trustchain_android.Util.Key
import nl.tudelft.cs4160.trustchain_android.database.TrustChainDBHelper

class ChainExplorerActivity : AppCompatActivity() {
    lateinit internal var dbHelper: TrustChainDBHelper


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chain_explorer)

        // Create a progress bar to display while the list loads
        val progressBar = ProgressBar(this)
        progressBar.layoutParams = LinearLayout.LayoutParams(GridLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT, CENTER.toFloat())
        progressBar.isIndeterminate = true
        blocksList.emptyView = progressBar

        // Must add the progress bar to the root of the layout
        val content = findViewById<ViewGroup>(android.R.id.content)
        content.addView(progressBar)

        init()
    }

    private fun init() {
        dbHelper = TrustChainDBHelper(this)
        val kp = Key.loadKeys(applicationContext)
        try {
            blocksList.adapter = ChainExplorerAdapter(this, dbHelper.allBlocks, kp!!.public.encoded)
        } catch (e: Exception) {
            e.printStackTrace()
        }

        blocksList.onItemClickListener = AdapterView.OnItemClickListener { parent, view, position, id ->
            // Expand the item when it is clicked
            if (view.expandedItem.visibility == View.GONE) {
                view.expandedItem.visibility = View.VISIBLE
                Log.v(TAG, "Item height: " + view.expandedItem.height)
                view.expandArrow.setImageDrawable(getDrawable(R.drawable.ic_expand_less_black_24dp))
            } else {
                view.expandedItem.visibility = View.GONE
                view.expandArrow.setImageDrawable(getDrawable(R.drawable.ic_expand_more_black_24dp))
            }
        }
    }

    companion object {

        internal val TAG = "ChainExplorerActivity"
    }

}
